[Candidate Parameters] React Revamp (Redmine 10761)

stellarxo opened the pull request:
TOTAL REACTIFICATION !!!!

alisterdev commented on modules/candidate_parameters/ajax/formHandler.php:
```diff
@@ -0,0 +1,309 @@
+    if ($tab == "candidateInfo") {
+        editCandInfoFields();
+    }
+    else if ($tab == "probandInfo") {
```
This is pretty weird if/else formatting. PHPCS would definitely help here

alisterdev commented on modules/candidate_parameters/ajax/formHandler.php:
```diff
@@ -0,0 +1,309 @@
+}
+
+function editCandInfoFields() {
+    $db   =& Database::singleton();
```
This block of code (line 30-35) is reused in all functions. Should probably go outside (e.g after line 4)

alisterdev commented on modules/candidate_parameters/ajax/getData.php:
```diff
@@ -0,0 +1,353 @@
+    foreach ($unformattedComments as $comment) {
+        $commentString = '';
+        $commentString .= '[' . $comment['data_entry_date'] . '] ';
+        $commentString .= '<i>Updated by ' . $comment['entry_staff'] . '. </i>';
```
The HTML (and static text) should be somewhere in react/template not in PHP. (This is especially disliked by @driusan 💀 )

alisterdev commented on modules/candidate_parameters/ajax/getData.php:
```diff
@@ -0,0 +1,353 @@
+        foreach ($unformattedComments as $comment) {
+            $commentString = '';
+            $commentString .= '[' . $comment['data_entry_date'] . '] ';
+            $commentString .= '<i>Updated by ' . $comment['entry_staff'] . '. </i>';
```
#2 The HTML (and static text) should be somewhere in react/template not in PHP.

alisterdev commented on modules/candidate_parameters/jsx/candidateInfo.js:
```diff
@@ -0,0 +1,298 @@
+        switch (extraParameters[key2].Type.substring(0, 3)) {
+          case "enu":
+            var types = extraParameters[key2].Type.substring(5);
+            types = types.slice(0, -1);
```
Would be nice to explain what kind of magic you're pulling out here

alisterdev commented:
Consent status tab breaks for me with this error

```bash
[Tue Sep 13 17:20:47.401598 2016] [:error] [pid 28799] [client xxx] PHP Warning:  PDOStatement::execute(): SQLSTATE[42S22]: Column not found: 1054 Unknown column 's' in 'field list' in /var/www/loris/php/libraries/Database.class.inc on line 688, referer: https://xxx.loris.ca/candidate_parameters/?candID=937726&identifier=937726
[Tue Sep 13 17:20:47.402173 2016] [:error] [pid 28799] [client xxx] Could not execute SELECT `s` FROM participant_status WHERE CandID=:candid LIMIT 1. Stack trace#0 /var/www/loris/php/libraries/Database.class.inc(716): Database->execute(Object(PDOStatement), Array)\n#1 /var/www/loris/php/libraries/Database.class.inc(733): Database->pselect('SELECT `s` FROM...', Array)\n#2 /var/www/loris/php/libraries/Database.class.inc(751): Database->pselectRow('SELECT `s` FROM...', Array)\n#3 /var/www/loris/modules/candidate_parameters/ajax/getData.php(306): Database->pselectOne('SELECT `s` FROM...', Array)\n#4 /var/www/loris/modules/candidate_parameters/ajax/getData.php(18): getConsentStatusFields()\n#5 /var/www/loris/htdocs/AjaxHelper.php(92): require('/var/www/loris/...')\n#6 {main}, referer: https://xxx.loris.ca/candidate_parameters/?candID=937726&identifier=937726
[Tue Sep 13 17:20:47.403000 2016] [:error] [pid 28799] [client xxx] PHP Fatal error:  Uncaught exception 'DatabaseException' with message 'Unknown column 's' in 'field list'' in /var/www/loris/php/libraries/Database.class.inc:691\nStack trace:\n#0 /var/www/loris/php/libraries/Database.class.inc(716): Database->execute(Object(PDOStatement), Array)\n#1 /var/www/loris/php/libraries/Database.class.inc(733): Database->pselect('SELECT `s` FROM...', Array)\n#2 /var/www/loris/php/libraries/Database.class.inc(751): Database->pselectRow('SELECT `s` FROM...', Array)\n#3 /var/www/loris/modules/candidate_parameters/ajax/getData.php(306): Database->pselectOne('SELECT `s` FROM...', Array)\n#4 /var/www/loris/modules/candidate_parameters/ajax/getData.php(18): getConsentStatusFields()\n#5 /var/www/loris/htdocs/AjaxHelper.php(92): require('/var/www/loris/...')\n#6 {main}\n  thrown in /var/www/loris/php/libraries/Database.class.inc on line 691, referer: https://xxx.loris.ca/candidate_parameters/?candID=937726&identifier=937726
```

stellarxo commented:
@alisterdev do you have database columns for all your consent types?

alisterdev commented:
@stellarxo I don't know, that's what I have in `participant_status` table
 
![image](https://cloud.githubusercontent.com/assets/6627543/18514936/7ba3b1f2-7a61-11e6-881b-dc8b6f0d935a.png)

stellarxo commented:
@alisterdev what is in your config.xml under study consent

xlecours commented:
@stellarxo I don't see how someone can set a Reason Specify in the participant status. I think that the available options should be the same than for Participant Status according to the foreign key 
`participant_suboptions integer DEFAULT NULL REFERENCES participant_status_options(ID),`

stellarxo commented:
@xlecours it's required only for status Inactive/Incomplete

stellarxo commented:
@xlecours I changed it so that the field is displayed and disabled if status != inactive or incomplete

stellarxo commented:
@alisterdev your config.xml should look like this:

<ConsentModule>
            <useConsent>true</useConsent>
            <Consent>
                <name>study_consent</name>
                <label>Consent to Study</label>
            </Consent>
            <Consent>
                <name>draw_blood_consent</name>
                <label>Consent to Draw Blood</label>
            </Consent>
</ConsentModule>

alisterdev commented on modules/candidate_parameters/js/familyInfo.js:
```diff
@@ -0,0 +1,303 @@
+    formData.append('familyDCCID', familyMemberID);
+    for (var field in familyMembers) {
+      if (familyMembers.hasOwnProperty(field)) {
+        console.log(familyMembers[field]);
```
Remove console.log

stellarxo replied:
c7266ea

alisterdev commented on modules/candidate_parameters/templates/form_candidate_parameters.tpl:
```diff
@@ -1,150 +1,123 @@
+    }
+    else {
+        $('#proband-info-tab').hide();
+        console.log(loris.config('useProband'));
```
Remove console.log

stellarxo replied:
c7266ea

alisterdev commented on modules/candidate_parameters/jsx/familyInfo.js:
```diff
@@ -0,0 +1,347 @@
+      formData.append('familyDCCID', familyMemberID);
+      for (var field in familyMembers) {
+        if (familyMembers.hasOwnProperty(field)) {
+          console.log(familyMembers[field]);
```
Remove console.log

stellarxo replied:
c7266ea

alisterdev commented on modules/candidate_parameters/jsx/candidateInfo.js:
```diff
@@ -0,0 +1,324 @@
+                        );
+          },
+          error: function(err) {
+            var errorMessage = JSON.parse(err.responseText).message;
```
If a random error occurs on the server, `err.responseText` is set to `""` and this line generates an error.
So you should probably check for empty and/or fallback on statusText

stellarxo replied:
Should probably be fixed in the media module as well

stellarxo replied:
e75fb3e

alisterdev commented on modules/candidate_parameters/ajax/formHandler.php:
```diff
@@ -0,0 +1,397 @@
+ */
+function editCandInfoFields()
+{
+    if (!$this->user->hasPermission('candidate_parameter_edit')) {
```
This PHP file is not a class (just a script), so `$this` always evaluates to `null`. I'm surprised that you got it to work this way, but everything crashes for me until I remove all `$this` references

alisterdev replied:
You can technically declare:
```
$db   =& Database::singleton();
$user =& User::singleton();
```
on line 17 right after `$tab`, and then pass references to all methods.

Otherwise use `global` keyword (probably not best practice)

alisterdev commented on modules/candidate_parameters/jsx/participantStatus.js:
```diff
@@ -0,0 +1,294 @@
+                    label ="DCCID"
+                    text ={this.state.Data.candID}
+                />
+                <SelectElement
```
Doesn't get save or populate after page refresh

stellarxo replied:
This seems to work for me. Can you show me on Tuesday on your machine?

alisterdev commented on modules/candidate_parameters/jsx/consentStatus.js:
```diff
@@ -0,0 +1,406 @@
+                        onUserInput ={this.setFormData}
+                        ref ={consentDate}
+                        disabled ={disabled}
+                        required ={dateRequired[i]}
```
Error on save (even after updating `config.xml`

alisterdev replied:
```
[Thu Sep 22 12:25:51.500067 2016] [:error] [pid 9845] [client xxx.xxx.xxx:51667] PHP Fatal error:  Uncaught exception 'DatabaseException' with message 'Update statement did not execute successfully.' in /var/www/loris/php/libraries/Database.class.inc:469\nStack trace:\n#0 /var/www/loris/php/libraries/Database.class.inc(391): Database->_realupdate('participant_sta...', Array, Array, true)\n#1 /var/www/loris/modules/candidate_parameters/ajax/formHandler.php(397): Database->update('participant_sta...', Array, Array)\n#2 /var/www/loris/modules/candidate_parameters/ajax/formHandler.php(30): editConsentStatusFields()\n#3 /var/www/loris/htdocs/AjaxHelper.php(92): require('/var/www/loris/...')\n#4 {main}\n  thrown in /var/www/loris/php/libraries/Database.class.inc on line 469, referer: https://xxx.loris.ca/candidate_parameters/?candID=843813&identifier=843813
```

stellarxo replied:
82459dd

driusan commented:
@alisterdev Why did you add the "Needs Work" tag instead of doing a review and setting it to "request changes"?

(Is there a problem with the permissions for reviews?)

alisterdev commented:
@driusan Not used to the new workflow, I just chose "comment" instead of "request changes" in my review

stellarxo commented:
@driusan I addressed all the issues. How do I mark that changes requested have been made?

![image](https://cloud.githubusercontent.com/assets/11700445/18805488/8e8166ec-81dd-11e6-9a29-6f024957ba7d.png)

alisterdev commented:
@stellarxo you need to make @driusan approve this PR/remove "request changes". Otherwise we don't seem to have a good technique to signal that changes were made

driusan commented:
I think the only way that GitHub has to indicate changes are made is to comment saying so.

driusan commented:
I didn't actually review this, I just approved it since @alisterdev did and I had marked the review as request changes on his behalf..

alisterdev commented:
I will re-review new changes and post an appropriate review

alisterdev commented on modules/candidate_parameters/jsx/participantStatus.js:
```diff
@@ -0,0 +1,295 @@
+                name ="participantStatus"
+                onSubmit ={this.handleSubmit}
+                ref ="form"
+                class ="col-md-6"
```
When you are adding a column class to an element, make sure that it (the element) is wrapped in another element with a `row` class to maintain correct layout. In this case the root `<div>` should have class `row`.

alisterdev replied:
@stellarxo It doesn't work if a selected candidate is not already in `participant_status` table. How is this table populated?

stellarxo replied:
fixed

alisterdev commented on modules/candidate_parameters/jsx/candidateInfo.js:
```diff
@@ -0,0 +1,326 @@
+            var xhr = new window.XMLHttpRequest();
+            xhr.addEventListener(
+                            "progress",
+                            function(evt) {
```
Run ESLint for formatting plzzz

stellarxo replied:
fixed

alisterdev commented on modules/candidate_parameters/jsx/candidateInfo.js:
```diff
@@ -0,0 +1,326 @@
+                        name ="flagged_caveatemptor"
+                        options ={this.state.caveatOptions}
+                        value ={this.state.Data.flagged_caveatemptor}
+                        onUserInput ={this.setFormData}
```
You should probably reset unused fields, before hiding them on option change. Same with error messages.
Otherwise, you might end up with stuff like this:

Field supposed to only be required for true
![image](https://cloud.githubusercontent.com/assets/6627543/18886754/44402d90-84bf-11e6-964e-33857de6517d.png)

Field should not have a value for false
![image](https://cloud.githubusercontent.com/assets/6627543/18886794/71afcf06-84bf-11e6-9883-4837a5814f6d.png)

alisterdev replied:
The `other` field disappears, once true is selected, and never comes back even if false is reselected

stellarxo replied:
fixed

alisterdev commented:
There seem to be no way to edit final review message 😢

alisterdev commented:
Once candidate info works as expected, this could be merged

stellarxo commented on modules/candidate_parameters/js/candidateInfo.js:
```diff
@@ -0,0 +1,308 @@
+      otherDisabled = true;
+      otherRequired = false;
+      this.state.formData.flagged_reason = '';
+      this.state.formData.flagged_other = '';
```
@alisterdev is there a better way to do this? I tried setting to null and it saved '0' and the string literal 'null'

alisterdev replied:
@stellarxo I think your problem is that you don't initialize `this.state.formData` in `componentDidMount`. You should create formData object in success callback.
Example [here](https://github.com/aces/Loris/blob/17.0-dev/modules/media/jsx/editForm.js#L35)

So essentially you should move all the value calculating `logic` from render() to a separate method that is called in succes()

alisterdev replied:
Also, to answer your question: 
1. You should use this.setState() to set the state
2. You can't set the state inside render as it results in 'infinite loop'

alisterdev replied:
Take the line 114-115 and move it inside setForm() after formData declaration as such:

```
if (formElement === "flagged_caveatemptor" && value === false) {
        formData['flagged_reason'] = '';
        formData['flagged_other'] = '';
      }
```

alisterdev replied:
Add this in `success` callback and use `this.state.formData.xxxx` for all the values in form elements

```
var formData = {
              flagged_caveatemptor: data.flagged_caveatemptor,
              caveatReasonOptions: data.caveatReasonOptions,
              flagged_reason: data.flagged_reason
            };
```

alisterdev commented on modules/candidate_parameters/js/candidateInfo.js:
```diff
@@ -32,9 +32,9 @@ var CandidateInfo = React.createClass({
       success: function (data) {
         var formData = {
-          flagged_caveatemptor: data.flagged_caveatemptor,
+          flaggedCaveatemptor: data.flagged_caveatemptor,
```
@stellarxo Lol, you broke it again. There are plenty of references that use `this.state.formData.flagged_caveatemptor`, etc

alisterdev replied:
Particularly, `value` of selectelements, text fields, etc

stellarxo replied:
@alisterdev can you test again?

alisterdev commented on modules/candidate_parameters/js/candidateInfo.js:
```diff
@@ -55,7 +55,7 @@ var CandidateInfo = React.createClass({
 
     // Reset 'reason' field
-    if (formElement === "flagged_caveatemptor" && value === "false") {
+    if (formElement === "flaggedCaveatemptor" && value === "false") {
```
this doesnt need to be changed

stellarxo replied:
fixed

alisterdev replied:
In fact, it does, but you would also need to change all names and references of selects to be camelCase

alisterdev commented on modules/candidate_parameters/js/candidateInfo.js:
```diff
@@ -64,7 +64,7 @@ var CandidateInfo = React.createClass({
 
     // Reset 'other' field
-    if (formElement === "flagged_reason" && value !== "2") {
+    if (formElement === "flaggedReason" && value !== "2") {
```
same

stellarxo replied:
fixed

driusan commented on modules/candidate_parameters/ajax/formHandler.php:
```diff
@@ -0,0 +1,417 @@
+    } else if ($tab == "participantStatus") {
+        editParticipantStatusFields($db, $user);
+    } else if ($tab == "consentStatus") {
+        editConsentStatusFields($db, $user);
```
I think this would be cleaner if each tab had it's own script instead of mixing everything together into one (but I don't think that's worth blocking the merge for.)
